xmtp_db/src/sql_key_store/transactions.rs

```diff
@@ -0,0 +1,186 @@
+use super::*;
+use crate::{DbConnection, TransactionGuard};
+use diesel::connection::LoadConnection;
+use diesel::migration::MigrationConnection;
+use diesel::sqlite::Sqlite;
+use diesel_migrations::MigrationHarness;
+use std::cell::RefCell;
+
+/// wrapper around a mutable connection (&mut SqliteConnection)
+/// Requires that all execution/transaction happens in one thread on one connection.
+/// This connection _must only_ be created from starting a transaction
+pub struct MutableTransactionConnection<'a, C> {
+    // we cannot avoid interior mutability here
+    // because raw_query methods require &self, as do MlsStorage trait methods.
+    // Since we no longer have async transactions, once a transaction is started
+    // we can ensure it occurs all on one thread.
+    pub(crate) conn: RefCell<&'a mut C>,
+}
+
+impl<'a, C> MutableTransactionConnection<'a, C> {
+    pub fn new(conn: &'a mut C) -> Self {
+        Self {
+            conn: RefCell::new(conn),
+        }
+    }
+}
+
+impl<'a, C> ConnectionExt for MutableTransactionConnection<'a, C>
+where
+    C: diesel::Connection<Backend = Sqlite>
+        + diesel::connection::SimpleConnection
+        + LoadConnection
+        + MigrationConnection
+        + MigrationHarness<<C as diesel::Connection>::Backend>
+        + crate::MlsKeyStore
+        + Send,
+{
+    type Connection = C;
+
+    fn start_transaction(&self) -> Result<TransactionGuard, crate::ConnectionError> {
+        Err(crate::ConnectionError::Database(
+            diesel::result::Error::AlreadyInTransaction,
+        ))
+    }
+
+    fn raw_query_read<T, F>(&self, fun: F) -> Result<T, crate::ConnectionError>
+    where
+        F: FnOnce(&mut Self::Connection) -> Result<T, diesel::result::Error>,
+        Self: Sized,
+    {
+        let mut conn = self.conn.borrow_mut();
+        fun(&mut conn).map_err(crate::ConnectionError::from)
+    }
+
+    fn raw_query_write<T, F>(&self, fun: F) -> Result<T, crate::ConnectionError>
+    where
+        F: FnOnce(&mut Self::Connection) -> Result<T, diesel::result::Error>,
+        Self: Sized,
+    {
+        let mut conn = self.conn.borrow_mut();
+        fun(&mut conn).map_err(crate::ConnectionError::from)
+    }
+
+    fn is_in_transaction(&self) -> bool {
+        true
+    }
+
+    fn disconnect(&self) -> Result<(), crate::ConnectionError> {
+        // generally does not make sense
```

Comment:
this generally does not make sense b/c the transaction mutex would need to close before we drop the db. The way we disconnect is dropping all references to sqlite connections that are held in the pool/mutexes, which we can't do from within a transaction, where we only have a &mut T

Reply:
should probably change this to an Error